setup/config/config_publickey.go

```diff
@@ -0,0 +1,92 @@
+package config
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"regexp"
+
+	"github.com/matrix-org/dendrite/clientapi/auth/authtypes"
+)
+
+var nonceByteLength = 32
+var minNonceCharacterLength = 8
+
+type AuthParams interface {
+	GetParams() interface{}
+	GetNonce() string
+}
+
+type EthereumAuthParams struct {
+	Version  uint   `json:"version"`
+	ChainIDs []int  `json:"chain_ids"`
+	Nonce    string `json:"nonce"`
+}
+
+func (p EthereumAuthParams) GetParams() interface{} {
+	copyP := p
+	copyP.ChainIDs = make([]int, len(p.ChainIDs))
+	copy(copyP.ChainIDs, p.ChainIDs)
+	copyP.Nonce = newNonce()
+	return copyP
+}
+
+func (p EthereumAuthParams) GetNonce() string {
+	return p.Nonce
+}
+
+type ethereumAuthConfig struct {
+	Enabled  bool  `yaml:"enabled"`
+	Version  uint  `yaml:"version"`
+	ChainIDs []int `yaml:"chain_ids"`
+}
+
+type publicKeyAuthentication struct {
+	Ethereum ethereumAuthConfig `yaml:"ethereum"`
+}
+
+func (pk *publicKeyAuthentication) Enabled() bool {
+	return pk.Ethereum.Enabled
+}
+
+func (pk *publicKeyAuthentication) GetPublicKeyRegistrationFlows() []authtypes.Flow {
+	var flows []authtypes.Flow
+	if pk.Ethereum.Enabled {
+		flows = append(flows, authtypes.Flow{Stages: []authtypes.LoginType{authtypes.LoginTypePublicKeyEthereum}})
+	}
+
+	return flows
+}
+
+func (pk *publicKeyAuthentication) GetPublicKeyRegistrationParams() map[string]interface{} {
+	params := make(map[string]interface{})
+	if pk.Ethereum.Enabled {
+		p := EthereumAuthParams{
+			Version:  pk.Ethereum.Version,
+			ChainIDs: pk.Ethereum.ChainIDs,
+			Nonce:    "",
+		}
+		params[authtypes.LoginTypePublicKeyEthereum] = p
+	}
+
+	return params
+}
+
+var regexpNotAlphaDigit = regexp.MustCompile("[^a-zA-Z0-9]+")
+
+func newNonce() string {
+	nonce := ""
+
+	for len(nonce) < minNonceCharacterLength {
```

Comment:
Is this loop going to run until a nonce is found that meets the regex? Wouldn't it be better to keep appending to the existing string until it's 8 alphanum?

Reply:
No it strips out any non-alphanumeric characters. If the remaining length > 8, it exits the loop immediately. In most cases, it should only run 1 time. Because the generated nonce is 32 chars long. 4 times the required min length. So stripping out any non-alphanumeric characters should still meet the min length.